File: delta/src/features/persist/kv.ts
// Copyright 2023 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { AsyncKV } from "@synnaxlabs/x";
import { invoke } from "@tauri-apps/api";

enum KVCommand {
  Get = "get",
  Set = "set",
  Delete = "delete",
}

interface KVResponse {
  value: string;
  error: string;
}

interface KVRequest {
  command: KVCommand;
  key: string;
  value: string;
}

/**
 * TauriKV an implementation of AsyncKV that communicates with a rust key-value
 * store running on the backend.
 */
export class TauriKV<V> implements AsyncKV<string, V> {
  async get(key: string): Promise<V | null> {
    try {
      return await this.exec({ command: KVCommand.Get, key, value: "" });
    } catch (err) {
      if ((err as Error).message === "Key not found") return null;
      throw err;
    }
  }

  async set(key: string, value: V): Promise<void> {
    await this.exec({
      command: KVCommand.Set,
      key,
      value: JSON.stringify(value),
    });
  }

  async delete(key: string): Promise<void> {
    await this.exec({ command: KVCommand.Delete, key, value: "" });
  }

  private async exec<V = null>(request: KVRequest): Promise<V | null> {
    const res: KVResponse = await invoke("kv_exec", { request });
    if (res.error.length > 0) throw new Error(res.error);
    if (res.value.length === 0) return null;
    return JSON.parse(res.value);
  }
}

File: synnax/pkg/ranger/kv.go
// Copyright 2023 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package ranger

import (
	"github.com/google/uuid"
	"github.com/synnaxlabs/x/gorp"
)

type KVPair struct {
	Range uuid.UUID `json:"range" msgpack:"range"`
	Key   string    `json:"key" msgpack:"key"`
	Value string    `json:"value" msgpack:"value"`
}

var _ gorp.Entry[string] = KVPair{}

// GorpKey implements gorp.Entry.
func (k KVPair) GorpKey() string { return k.Range.String() + "<--->" + k.Key }

// SetOptions implements gorp.Entry.
func (k KVPair) SetOptions() []interface{} { return nil }
